### ISSUE TITLE
Fast-LoRA autograd: fused LoRA MLP block (gate/up/down around SwiGLU) with a hand-written backward

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#837]
Part of #792 (fast-LoRA autograd tracker). The flag, validators, tolerance rule, NF4 rule, checkpointing and streaming criteria live there; this issue is the MLP math and its own tests.

## Why

HF's Llama MLP (transformers 5.17.0, `models/llama/modeling_llama.py:175`) is

```

down_proj(act_fn(gate_proj(x)) * up_proj(x))

```

With peft LoRA on all three projections, generic autograd saves, per block (counted from the op derivatives; the tracker's saved-tensor accounting confirms or corrects this):

- `X` — for gate's and up's `lora_A` (one storage at dropout 0)

- `G = gate(X)` — for the silu backward

- `silu(G)` and `U = up(X)` — for the product's backward

- `M = silu(G) * U` — for down's `lora_A`

- three `[N, r]` intermediates — for the `lora_B` weight gradients

That is `X` plus four `[N, d_ff]` tensors. With `use_liger`, `LigerSiLUMulFunction` saves `(G, U)` and recomputes the activation (`liger_kernel/ops/swiglu.py:272`), which removes `silu(G)` — but `M` is still saved, by the peft-wrapped `down_proj` that Liger calls (`liger_kernel/transformers/swiglu.py:35`).

A hand-written backward keeps `X`, `G`, `U` and recomputes `silu(G)` and `M`, which is elementwise. Unsloth's `LoRA_MLP` saves the same set (`unsloth/kernels/fast_lora.py:107`: the six adapter matrices, `X`, `e`, `g`).

## Math

peft/torch convention: `W: [out, in]`, `A: [r, in]`, `B: [out, r]`, `X: [N, d]`, `s = alpha / r` (or `alpha / sqrt(r)` under rsLoRA, `peft/tuners/lora/layer.py:275-277`), dropout 0.

Forward:

```

H_gu = X @ cat(A_g, A_u)^T              # ONE GEMM, [N, r_g + r_u]; split -> H_g, H_u

G    = X @ W_g^T + s_g * H_g @ B_g^T    # [N, f]

U    = X @ W_u^T + s_u * H_u @ B_u^T    # [N, f]

M    = silu(G) * U                      # [N, f]

H_d  = M @ A_d^T                        # [N, r_d]

Y    = M @ W_d^T + s_d * H_d @ B_d^T    # [N, d]

```

Saved: `X`, `G`, `U`, the A/B parameters, and the base weights (or packed NF4 + quant-state tensors) through `save_for_backward`. The `[N, r]` H's are either saved or recomputed; pick one and justify it with the measurement.

Backward, given `dY: [N, d]`, `sig = sigmoid(G)`:

```

M     = silu(G) * U                            # recompute

H_d   = M @ A_d^T                              # recompute (or saved)

dB_d  = s_d * dY^T @ H_d

dH_d  = s_d * dY @ B_d                         # [N, r_d]

dA_d  = dH_d^T @ M

dM    = dY @ W_d + dH_d @ A_d                  # accumulated into one buffer

dU    = dM * silu(G)

dG    = dM * U * sig * (1 + G * (1 - sig))     # silu'(G)

dB_g  = s_g * dG^T @ H_g ;  dB_u = s_u * dU^T @ H_u

dH_gu = cat(s_g * dG @ B_g, s_u * dU @ B_u)    # [N, r_g + r_u]

dA_gu = dH_gu^T @ X                            # ONE GEMM; split -> dA_g, dA_u

dX    = dG @ W_g + dU @ W_u + dH_gu @ cat(A_g, A_u)

```

For NF4, `W_g`, `W_u`, `W_d` are dequantised in backward one at a time, so the transient is one dense `[f, d]` matrix, not three.

## Scope

In: MLPs with `gate_proj` / `up_proj` / `down_proj` children and a SiLU activation — Llama, Mistral, Qwen2, Qwen3. Detect by structure and by `config.hidden_act` in `{"silu", "swish"}` (`transformers/activations.py:343-345`), not by model name. Any subset of the three may be adapted; an unadapted projection is base-only and its A/B terms drop out. Note Soup's `target_modules: auto` on Llama adapts none of them (peft default `["q_proj", "v_proj"]`), so this path engages only when the user lists them — the setup print must say how many MLP blocks were patched.

Out: GeGLU (Gemma's tanh-GELU has a different derivative — follow-up), fused `gate_up_proj` (Phi-3), MoE experts, any other activation (fall back to the peft path and say so).

## Tests (in addition to the tracker's shared set)

- [ ] gradcheck, float64, CPU: `N=5, d=8, f=12, r=2` for all three, w.r.t. X and all six adapter matrices; repeated with each projection in turn unadapted, and with unequal ranks via `rank_pattern`.

- [ ] Saturation: a case with `G` containing values around +/-20, so `sig` at 0 and 1 is exercised in bf16.

- [ ] Parity vs an unpatched `LlamaMLP` + peft: `Y` (forward) and `dX`, every `dA`, every `dB` (backward) reported separately, fp32 and bf16, plain and NF4.

- [ ] Saved-tensor bytes at the Llama-3.1-8B MLP shape (`N=2048, d=4096, f=14336, r=64`): fast vs plain vs plain + `use_liger`. The count above predicts two `[N, f]` tensors fewer than plain and one fewer than Liger; the measured number is what gets written down.

- [ ] Micro-benchmark, one MLP block at that shape, bf16 and NF4: median forward+backward time (CUDA events, warm-up, repeats) and `torch.cuda.max_memory_allocated`, fast vs plain vs plain + Liger, posted with card, driver, SM clock and versions.

- [ ] `hidden_act` other than silu falls back with a printed reason.

## Hardware needed

CPU for gradcheck and fp32 parity. Any CUDA GPU for bf16/NF4 parity and the micro-benchmark; one bf16 MLP block at the 8B shape is `3 x 4096 x 14336 x 2` bytes, about 352 MB of weights (arithmetic, not a measurement), so an 8 GB card is enough.

## Honest limits

- Under gradient checkpointing the saving lowers a per-layer transient, not the stored-activation total (see tracker).

- Recomputing `silu(G)` and `M` trades elementwise work for memory; whether that pays at bf16 on a given card is what the micro-benchmark answers.

- bf16 parity is tolerance-based, not bit-exact.

---

**Related, filed the same day:**

- #792 — Fast-LoRA autograd, correctness half: hand-written backward for the LoRA projections, behind training.fast_...

**Comment by Amix29** [comment]
I’ll take this. I’ll implement the fused LoRA SwiGLU MLP path as the correctness-first slice from #792: structural SiLU/swish detection, all/partial adapted gate/up/down projections, unequal ranks, a hand-written autograd backward, CPU float64 gradcheck and fp32 parity first, then Apple-Silicon/MPS measurements where the stack supports them. I’ll keep CUDA-specific benchmark claims out unless they are actually measured, and I’ll make fallback behavior explicit for unsupported activations/shapes.

**Comment by Amix29** [comment]
Implementation is published as draft PR #1107 from Amix29. It is intentionally stacked on #1087/#839 until the shared Fast-LoRA helper lands. Local Apple M4 Max validation: 33 passed / 2 CUDA-only skips across #837+#839; float64 gradcheck, real LlamaMLP fp32 parity, partial/unequal-rank adapters, bf16 MPS saturation, and saved-tensor reduction 33,376 -> 18,016 bytes. No CUDA performance claim.
